core/dive.cpp

```diff
@@ -367,91 +367,117 @@ static bool has_unknown_used_cylinders(const struct dive &dive, const struct div
-		 */
-		for (size_t i = 0; i < dive->cylinders.size(); i++) {
-			if (used_cylinders[i]) {
-				mean[i] = dc->meandepth.mm;
-				duration[i] = dc->duration.seconds;
-			}
-		}
-
-		return;
-	}
-	if (dc->samples.empty())
-		fake_dc(dc);
-
-	gasmix_loop loop(*dive, *dc);
-	std::vector<int> depthtime(dive->cylinders.size(), 0);
-	int lasttime = 0;
-	int lastdepth = 0;
-	int last_cylinder_index = -1;
-	std::pair<int, int> gaschange_event;
-	for (auto it = dc->samples.begin(); it != dc->samples.end(); ++it) {
-		int32_t time = it->time.seconds;
-		int depth = it->depth.mm;
-
-		/* Make sure to move the event past 'lasttime' */
-		gaschange_event = loop.cylinder_index_at(lasttime);
-
-		/* Do we need to fake a midway sample? */
-		if (last_cylinder_index >= 0 && last_cylinder_index != gaschange_event.first) {
-			int newdepth = interpolate(lastdepth, depth, gaschange_event.second - lasttime, time - lasttime);
-			if (newdepth > SURFACE_THRESHOLD || lastdepth > SURFACE_THRESHOLD) {
-				duration[gaschange_event.first] += gaschange_event.second - lasttime;
-				depthtime[gaschange_event.first] += (gaschange_event.second - lasttime) * (newdepth + lastdepth) / 2;
-			}
-
-			lasttime = gaschange_event.second;
-			lastdepth = newdepth;
-		}
-
-		/* We ignore segments at the surface */
-		if (depth > SURFACE_THRESHOLD || lastdepth > SURFACE_THRESHOLD) {
-			duration[gaschange_event.first] += time - lasttime;
-			depthtime[gaschange_event.first] += (time - lasttime) * (depth + lastdepth) / 2;
-		}
-
-		lastdepth = depth;
-		lasttime = time;
-		last_cylinder_index = gaschange_event.first;
-	}
-	for (size_t i = 0; i < dive->cylinders.size(); i++) {
-		if (duration[i])
-			mean[i] = (depthtime[i] + duration[i] / 2) / duration[i];
-	}
+void per_cylinder_mean_depth(const struct dive *dive, struct divecomputer *dc, int *mean, int *duration) {
+    // Return early if the dive computer is null or there are no cylinders
+    if (!dc || dive->cylinders.empty())
+        return;
+
+    // Initialize mean and duration arrays to zero for each cylinder
+    initialize_cylinder_data(dive, mean, duration);
+
+    // Determine which cylinders were used during the dive
+    auto used_cylinders = std::make_unique<bool[]>(dive->cylinders.size());
+    int num_used_cylinders = get_cylinder_used(dive, used_cylinders.get());
+
+    // Handle cases where cylinder usage is unknown
+    if (handle_unknown_cylinders(*dive, dc, used_cylinders.get(), num_used_cylinders, mean, duration))
+        return;
+
+    // Simulate samples if none exist in the dive computer
+    if (dc->samples.empty())
+        fake_dc(dc);
+
+    // Calculate depth and duration for each cylinder
+    calculate_depth_and_duration(dive, dc, mean, duration);
+}
+
+// Helper function: Initializes mean and duration arrays for all cylinders
+void initialize_cylinder_data(const struct dive *dive, int *mean, int *duration) {
+    for (size_t i = 0; i < dive->cylinders.size(); ++i) {
+        mean[i] = 0;
```

Comment:
I agree with @atdotde here that this is going a bit over the top - refactoring into multiple functions (that are each used only once) doesn't actually make this easier to read or understand. And thre is no reason that initialisation and finalisation of the function should not happen in the function body.

Reply:
Yes, this doesn't even decrease the overall depth of indentation, does it? What is the point?